// derby-examples/sink/lib/app/index.js
// Generated by CoffeeScript 1.3.1
var app, get, ready, render, _ref;

_ref = app = require('derby').createApp(module), get = _ref.get, ready = _ref.ready;

render = require('./shared').render;

require('./home');

require('./live-css');

require('./table');

require('./leaderboard');

require('./bindings-bench');

['get', 'post', 'put', 'del'].forEach(function(method) {
  return app[method]('/submit', function(page, model, _arg) {
    var args, body, query;
    body = _arg.body, query = _arg.query;
    args = JSON.stringify({
      method: method,
      body: body,
      query: query
    }, null, '  ');
    return render(page, 'submit', {
      args: args
    });
  });
});

get('/error', function() {
  throw new Error(500);
});

get('/back', function(page) {
  return page.redirect('back');
});

ready(function(model) {
  model.set('_showReconnect', true);
  exports.connect = function() {
    model.set('_showReconnect', false);
    setTimeout((function() {
      return model.set('_showReconnect', true);
    }), 1000);
    return model.socket.socket.connect();
  };
  return exports.reload = function() {
    return window.location.reload();
  };
});
